slides/week3/week3.org

```diff
@@ -2,66 +2,158 @@
 
-* Common containers
+* Vectors
 #+BEGIN_NOTES
-Note that in most languages, sets and maps are unorded. C++ also has unordered sets and maps.
+initializer lists require c++11 for vectors.
 #+END_NOTES
-| =array= | a static-length ordered list of objects |
-| =vector= | a dynamic-length ordered list of objects |
-| =set= | an ordered group of unique objects |
-| =unordered_set= | an unordered group of unique objects |
-| =map= | an ordered group of key-value pairs, with unique keys |
-| =unordered_map= | an unordered group of key-value pairs, with unique keys |
+- an object that can contain a variable number of other objects
+- you can keep adding elements to the vector
+- order of insertion in maintained
+#+BEGIN_SRC C++
+  #include <vector>
 
-* Arrays
-| =at= | returns the value at a specific index |
-| =front= | returns the first item |
-| =back= | returns the last item |
-| =size= | returns the number of elements |
+  vector<int> name = {1, 2, 3};
+#+END_SRC
 
-* Vectors
+* Vector methods
 #+BEGIN_NOTES
 Vectors can grow as items are added, whereas arrays are static.
 #+END_NOTES
-- Everything from the last slide, and...
-| =clear= | removes all items |
-| =insert= | inserts an item at a specific index |
-| =push_back= | adds a given element to the end |
-| =pop_back= | removes the last element |
+- Everything from the arrays slide, and...
+| =clear=     | removes all items                   |
+| =insert=    | inserts an item at a specific index |
+| =push_back= | adds a given element to the end     |
+| =pop_back=  | removes the last element            |
 
 * Iterators
 - An object that lets you access and modify objects in a container
-- Depending on their type, they will let you read or write data in the container, and move forwards, backwards, or arbitrarily
+- Depending on their type, they will let you read or write data in the container, and move forwards, backwards, or both
 - The containers library includes the most common ones you'll need, but you can also write your own
 
-* Algorithms
+* Iterator methods
 #+BEGIN_NOTES
-All of these can be restricted to a range in the container using iterators. All of these are in <algorithm> except for accumulate and iota, which are in <numeric>.
+picture upcoming. mention that the method should be called on the containing object
 #+END_NOTES
-| =sort= | sorts a container |
-| =count= | counts the number of items in a container that match a given item |
-| =find= | returns an iterator to the first element that matches a given item |
+| =*=      | gets the value at the current index                     |
+| =begin=  | starts at the *first* item that moves *up* in indexes   |
+| =end=    | starts at the *last* item that moves *up* in indexes    |
+| =rbegin= | starts at the *first* item that moves *down* in indexes |
+| =rend=   | starts at the *last* item that moves *down* in indexes  |
+
+* Moving iterators
+#+BEGIN_NOTES
+picture upcoming
+#+END_NOTES
+- =++= moves the iterator forwards
```

Comment:
With reverse iterators, won't this move them backwards?

Perhaps, 'increments the iterator'